reproman/interface/docker.shim

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/perl
+#emacs: -*- mode: shell-script; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: t -*-
+#ex: set sts=4 ts=4 sw=4 noet:
+#
+#
+# This script is a shim for the docker executable. The shim intercepts
+# calls to the docker executable, runs a image id discovery process, and
+# then proceeds with the original call to the original docker executable.
+#
+require 5.002;
+
+# Print debugging if docker debug switch is set
+my $DEBUG = 0;
+foreach(@ARGV) {
+    $DEBUG = 1 if ($_ eq '--debug' or $_ eq '-D');
+}
+
+# Print debug output
+# 
+# Parameters
+# ----------
+# $output
+#     string : Text to print to the screen.
+# 
+sub debug {
+    return if (!$DEBUG);
+    foreach my $line (@_) {
+        print "[DEBUG DOCKER SHIM] ${line}";
+        print "\n" if (!($line =~ /\n$/));
+    }
+}
+
+# Run commands in shell
+# 
+# Parameters
+# ----------
+# $command
+#     string : Command to execute in the shell
+# $default_message
+#     string : Message to send if nothing is found in command stderr
+#
+# Returns
+# -------
+# string : stdout of the executed command
+# 
+sub exec_handler {
+
+	my $command = shift;
+	my $default_message = shift;
+
+	my $output = qx/$command 2>&1/;
+	if ($? != 0) {
+		if ($output) {
+			die $output
+		}
+		else {
+  		die $default_message;
+		}
+	}
+	return $output
+}
+
+# Gather ENV info
+my $TRACERS_DIR = $ENV{"REPROMAN_TRACERS_DIR"};
+if (!$TRACERS_DIR) {
+	die "ENV variable REPROMAN_TRACERS_DIR not found";
+}
+debug("TRACERS_DIR: ${TRACERS_DIR}");
+my $EXTRA_TRACE_FILE = $ENV{"REPROMAN_EXTRA_TRACE_FILE"};
+if (!$EXTRA_TRACE_FILE) {
+	die "ENV variable REPROMAN_EXTRA_TRACE_FILE not found";
+}
+debug("EXTRA_TRACE_FILE: ${EXTRA_TRACE_FILE}");
+my $WAIT_SCRIPT = "wait.sh";
+
+# Locate original Docker executable (Most likely in /usr/bin)
+my $stdout = exec_handler("which -a docker", "Unable to execute system 'which' command");
+my @paths = split(/\s+/, $stdout);
+my $original_docker_exec = '';
+for my $path (@paths) {
+  $original_docker_exec = $path if $path ne $0;
+}
```

Comment:
I don't quite get what's going on here.  My guess is that we use `-a` because the first item should be our own shim, but, if there's more than one non-shim docker in the path, won't the above set the original one to the last one, which wouldn't be the one would take precedence in the unwrapped docker call?

Reply:
Yes, this is a problem. It should get the second item in the list